File: ios/Bridge/Handlers/StaticBarcodeScanningHandler.swift
import Foundation

#if MLKIT_BARCODE_SCANNING
  class StaticBarcodeScanningHandler: IStaticImageHandler {
    private var cachedUseCases: [String: RecognizeBarcodesUseCase] = [:]

    private func getRecognizeBarcodesUseCase(for options: BarcodeScanningOptions)
      -> RecognizeBarcodesUseCase
    {
      let key =
        "\(options.enableAllPotentialBarcodes)|\(options.formats.map(\.rawValue).joined(separator: ","))"

      if let useCase = cachedUseCases[key] {
        return useCase
      }

      let newUseCase = RecognizeBarcodesUseCase(
        imagePreprocessor: ImagePreprocessor(),
        recognitionService: BarcodeScanningServiceFactory.create(options: options)
      )
      cachedUseCases[key] = newUseCase
      return newUseCase
    }

    func process(
      path: String,
      options: [String: Any],
      resolver: @escaping PromiseResolver,
      rejecter: @escaping PromiseRejecter
    ) {
      Task {
        do {
          guard let fileURL = resolveFileURL(from: path) else {
            rejecter(
              "IMAGE_NOT_FOUND_ERROR",
              "Unsupported image URI: \(path)",
              nil
            )
            return
          }

          if fileURL.isFileURL
            && !FileManager.default.fileExists(atPath: fileURL.path)
          {
            rejecter(
              "IMAGE_NOT_FOUND_ERROR",
              "Image file not found at path: \(path)",
              nil
            )
            return
          }

          let imageOptions = parseImageOptions(options)
          let barcodeOptions = parseBarcodeOptions(options)
          let recognizeBarcodesUseCase = getRecognizeBarcodesUseCase(for: barcodeOptions)

          let result = try recognizeBarcodesUseCase.execute(
            imageFile: fileURL,
            imageOptions: imageOptions,
            barcodeOptions: barcodeOptions
          )

          resolver(BarcodeScanningSerializer.toReactNativeMap(result))
        } catch {
          switch error {
          case let nsError as NSError:
            if nsError.domain == NSCocoaErrorDomain
              && nsError.code == NSFileReadNoSuchFileError
            {
              rejecter("IMAGE_NOT_FOUND_ERROR", "Image file not found", error)
            } else {
              rejecter(
                "IMAGE_PROCESSING_FAILED_ERROR",
                error.localizedDescription,
                error
              )
            }
          default:
            rejecter(
              "IMAGE_PROCESSING_FAILED_ERROR",
              error.localizedDescription,
              error
            )
          }
        }
      }
    }

    private func parseImageOptions(_ options: [String: Any])
      -> ImagePreprocessingOptions
    {
      let invertColors = options["invertColors"] as? Bool ?? false
      let orientation = (options["orientation"] as? String).flatMap { DomainOrientation(string: $0) }
      let scaleFactor = CGFloat((options["scaleFactor"] as? NSNumber)?.doubleValue ?? 1.0)
      let roi = parseRegionOfInterest(options["roi"])

      return ImagePreprocessingOptions(
        invertColors: invertColors,
        scaleFactor: scaleFactor,
        orientation: orientation,
        roi: roi
      )
    }

    private func parseRegionOfInterest(_ value: Any?) -> DomainRegionOfInterest? {
      guard let dictionary = value as? [String: Any],
        let x = (dictionary["x"] as? NSNumber)?.doubleValue,
        let y = (dictionary["y"] as? NSNumber)?.doubleValue,
        let width = (dictionary["width"] as? NSNumber)?.doubleValue,
        let height = (dictionary["height"] as? NSNumber)?.doubleValue
      else {
        return nil
      }

      let unit: DomainRegionOfInterestUnit =
        (dictionary["unit"] as? String) == "pixel" ? .pixel : .normalized

      return DomainRegionOfInterest(
        x: CGFloat(x),
        y: CGFloat(y),
        width: CGFloat(width),
        height: CGFloat(height),
        unit: unit
      )
    }

    private func parseBarcodeOptions(_ options: [String: Any])
      -> BarcodeScanningOptions
    {
      return BarcodeScanningOptions(
        formats: BarcodeScanningOptionParser.parseFormats(options["formats"]),
        enableAllPotentialBarcodes:
          BarcodeScanningOptionParser
          .parseEnableAllPotentialBarcodes(options["enableAllPotentialBarcodes"])
      )
    }

    private func resolveFileURL(from path: String) -> URL? {
      if path.hasPrefix("file://") {
        return URL(string: path)
      }

      if let url = URL(string: path), url.scheme != nil {
        return url
      }

      return URL(fileURLWithPath: path)
    }
  }
#endif  // MLKIT_BARCODE_SCANNING

File: android/src/main/java/com/visioncameramlkit/bridge/handlers/StaticBarcodeScanningHandler.kt
package com.visioncameramlkit.bridge.handlers

import android.net.Uri
import androidx.core.net.toUri
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReadableMap
import com.visioncameramlkit.application.usecases.RecognizeBarcodesUseCase
import com.visioncameramlkit.bridge.parsers.BarcodeScanningOptionParser
import com.visioncameramlkit.domain.models.BarcodeScanningOptions
import com.visioncameramlkit.domain.models.ImagePreprocessingOptions
import com.visioncameramlkit.domain.models.Orientation
import com.visioncameramlkit.infrastructure.image.ImagePreprocessor
import com.visioncameramlkit.infrastructure.mlkit.factories.BarcodeScanningServiceFactory
import com.visioncameramlkit.infrastructure.serializers.BarcodeScanningSerializer
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.io.File
import java.io.FileOutputStream

class StaticBarcodeScanningHandler(
  private val reactContext: ReactApplicationContext,
) : IStaticImageHandler {
  private val cachedUseCases = mutableMapOf<String, RecognizeBarcodesUseCase>()

  private fun getRecognizeBarcodesUseCase(options: BarcodeScanningOptions): RecognizeBarcodesUseCase {
    val cacheKey =
      buildString {
        append(options.enableAllPotentialBarcodes)
        append("|")
        append(options.formats.joinToString(separator = ",") { it.name })
      }

    return cachedUseCases.getOrPut(cacheKey) {
      RecognizeBarcodesUseCase(
        ImagePreprocessor(),
        BarcodeScanningServiceFactory.create(options),
      )
    }
  }

  override fun process(
    path: String,
    options: ReadableMap,
    promise: Promise,
  ) {
    CoroutineScope(Dispatchers.IO).launch {
      var resolvedImage: ResolvedImage? = null
      try {
        resolvedImage = resolveImage(path)
        val imageFile = resolvedImage.file
        if (!imageFile.exists()) {
          promise.reject("IMAGE_NOT_FOUND_ERROR", "Image file not found at path: $path")
          return@launch
        }

        val imageOptions = parseImageOptions(options)
        val barcodeOptions = parseBarcodeOptions(options)
        val recognizeBarcodesUseCase = getRecognizeBarcodesUseCase(barcodeOptions)

        val result = recognizeBarcodesUseCase.execute(imageFile, imageOptions, barcodeOptions)
        val serializedResult = BarcodeScanningSerializer.toWritableMap(result)

        promise.resolve(serializedResult)
      } catch (
        @Suppress("TooGenericExceptionCaught") e: Exception,
      ) {
        when (e) {
          is IllegalArgumentException -> {
            promise.reject("INVALID_URI_ERROR", e.message)
          }

          is UnsupportedOperationException -> {
            promise.reject(
              "UNSUPPORTED_IMAGE_FORMAT_ERROR",
              e.message,
            )
          }

          else -> {
            promise.reject(
              "IMAGE_PROCESSING_FAILED_ERROR",
              e.message ?: "Barcode scanning failed",
            )
          }
        }
      } finally {
        val imageFile = resolvedImage?.file
        if (resolvedImage?.shouldDelete == true && imageFile != null && imageFile.exists()) {
          imageFile.delete()
        }
      }
    }
  }

  private data class ResolvedImage(
    val file: File,
    val shouldDelete: Boolean,
  )

  private fun resolveImage(path: String): ResolvedImage {
    val uri = path.toUri()
    return when (uri.scheme) {
      "content" -> ResolvedImage(copyContentToCache(uri), true)
      "file" -> ResolvedImage(File(uri.path ?: path), false)
      null -> ResolvedImage(File(path), false)
      else -> ResolvedImage(File(path), false)
    }
  }

  private fun copyContentToCache(uri: Uri): File {
    val cacheFile = File(reactContext.cacheDir, "mlkit_${System.currentTimeMillis()}.tmp")
    val inputStream =
      reactContext.contentResolver.openInputStream(uri)
        ?: throw IllegalArgumentException("Unable to open content URI: $uri")
    inputStream.use { input ->
      FileOutputStream(cacheFile).use { output ->
        input.copyTo(output)
      }
    }
    return cacheFile
  }

  private fun parseImageOptions(options: ReadableMap): ImagePreprocessingOptions =
    ImagePreprocessingOptions(
      invertColors = options.getOrDefault("invertColors", false),
      orientation = options.getString("orientation")?.let { parseOrientation(it) },
      scaleFactor = options.getOrDefault("scaleFactor", 1.0f),
    )

  private fun parseBarcodeOptions(options: ReadableMap): BarcodeScanningOptions =
    BarcodeScanningOptions(
      formats = BarcodeScanningOptionParser.parseFormats(options),
      enableAllPotentialBarcodes =
        BarcodeScanningOptionParser.parseEnableAllPotentialBarcodes(
          options,
        ),
    )

  private fun parseOrientation(orientation: String): Orientation? =
    when (orientation) {
      "portrait" -> Orientation.PORTRAIT
      "portrait-upside-down" -> Orientation.PORTRAIT_UPSIDE_DOWN
      "landscape-left" -> Orientation.LANDSCAPE_LEFT
      "landscape-right" -> Orientation.LANDSCAPE_RIGHT
      else -> null
    }

  private fun ReadableMap.getOrDefault(
    key: String,
    defaultValue: Boolean,
  ): Boolean {
    if (!hasKey(key) || isNull(key)) {
      return defaultValue
    }
    return getBoolean(key)
  }

  private fun ReadableMap.getOrDefault(
    key: String,
    defaultValue: Float,
  ): Float {
    if (!hasKey(key) || isNull(key)) {
      return defaultValue
    }
    return getDouble(key).toFloat()
  }
}
